Adds the ability to perform forward-split Cahn-Hilliard calculations with thermodiffusion and automatic differentiation

jah5759 opened the pull request:
Adds the ability to perform forward-split Cahn-Hilliard calculations with thermodiffusion and
  automatic differentiation
Includes
  - AD versions of the CHSplitChemicalPotential and CHSplitConcentration kernels
  - A new ADCHSoretMobility kernel, which accepts a thermal mobility parameter
  - Documentation and tests for the new objects

closes # 14396

Commit:
forward_split_cahn_hilliard_with_thermodiffusion_AD_#14396

Adds the ability to perform forward-split Cahn-Hilliard calculations with thermodiffusion and
  automatic differentiation
Includes
  - AD versions of the CHSplitChemicalPotential and CHSplitConcentration kernels
  - A new ADCHSoretMobility kernel, which accepts a thermal mobility parameter
  - Documentation and tests for the new objects

closes # 14396

jah5759 commented:
@tonkmr

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/433607/) on 40ce07a wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/14399/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/14399/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format e78fc3bf0a6252dac27ac669efcab34474f701a2`

dschwen commented:
`ADCHSplitChemicalPotential` will not get forward-mode derivatives from the `DerivativeParsedMaterial`!

dschwen reviewed (changes requested):
The tests need Jacobian tests, which will show you that your Jacobian will be wrong, because we do not yet have a dual number version of DerivativeParsedMaterial. I'm not in favor of adding this before we have a dual number parsed material. This is just asking for problems.

jah5759 commented:
@dschwen My application uses materials put together with AD behind the scenes instead of the parsed material system, but I cannot include these in the tests.  I could use constant material properties or `ADPiecewiseLinearInterpolationMaterial` in the tests to avoid the parsed material system for now.  I'll also look into the Jacobian tests.

jah5759 commented:
I replaced the material properties (mobilities) with constants and used `ADPiecewiseLinearInterpolationMaterial` to link the concentration and chemical potential within the AD system.  I added Jacobian tests and reduced the size of the tests to keep the timing down.

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/433835/) on da2d6b6 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/14399/site)

This comment will be updated on new commits.

dschwen commented:
Sounds good. In the meantime I started on parsed function AD support.

jah5759 commented:
Great, thanks again.

tonkmr commented:
@dschwen these changes are needed for work that is going on in BISON. Is there any way we can merge them now, since they do work and are tested without the parsed function AD support, and then just add tests using parsed functions with AD support once that is merged?

dschwen commented on modules/phase_field/src/kernels/ADCHSplitConcentration.C:
```diff
@@ -0,0 +1,35 @@
+ADReal
+ADCHSplitConcentration<compute_stage>::computeQpResidual()
+{
+  return _mobility[_qp] * _grad_mu[_qp] * _grad_test[_i][_qp];
```
Hm, we have a similar Kernel in `ADSplitCHCRes` let me review this more closely. I'm concerned we are creating parallel ways to do the same thing.

jah5759 replied:
I believe `ADSplitCHCRes` is intended to accept the chemical potential as an argument rather than the kernel variable itself.  It includes the following two lines:

`residual += -_w[_qp] * _test[_i][_qp];`
 `residual += _kappa[_qp] * _grad_u[_qp] * _grad_test[_i][_qp];`

So even if you applied `ADSplitCHCRes` with the kernel variable being the chemical potential, you'd get:
residual = -mu * test + kappa * grad_mu * grad_test
This is different than what we're proposing here:
residual = M * grad_mu * grad_test

The non-AD kernel that our kernel is based on is `CHSplitConcentration`, which has:

`const RealVectorValue a = _mobility[_qp] * _grad_mu[_qp];`
 `return _grad_test[_i][_qp] * a;`

This gives the same results as what we're proposing in AD form.  If we are creating a parallel calculation, is this kernel also parallel to `SplitCHCRes`?

dschwen commented on modules/phase_field/doc/content/source/kernels/ADCHSplitChemicalPotential.md:
```diff
@@ -0,0 +1,16 @@
+
+!syntax description /Kernels/ADCHSplitChemicalPotential
+
+Uses automatic differentiation to solve for chemical potential in a weak sense (mu-mu_prop=0).  Can
```
This is a parallel class to `ADSplitCHParsed` :-/

jah5759 replied:
`ADSplitCHParsed` takes the concentration variable and a material free energy.  It returns the material property holding the derivative of the free energy with respect to the concentration:

`_dFdc(getADMaterialProperty<Real>(derivativePropertyNameFirst(_f_name, _var.name())))`
...
`return _dFdc[_qp];`

Our application needs to accept a chemical potential variable and a material chemical potential.  This allows the material chemical potential to be supplied by an external solver (or other means) without ever explicitly defining the free energy.  

What we're proposing is an AD version of `CHSplitChemicalPotential`.  I don't think this provides a parallel calculation to `SplitCHParsed`.

dschwen replied:
Ah, ok, that seems reasonable. I'll go ahead and approve this for merging. At some point it would be nice to link the docs of the AD and non-AD versions and to add a `use_automatic_differentiation` option to the phase field actions.

dschwen replied:
Ugh, it even stated that in the description of the PR...

jah5759 replied:
Sounds good, thanks again!  I think having linked (or even combined) documentation for objects that have both standard and AD versions is a good idea.
